### PR TITLE
Add an option `end` to `find_in_batches`

### PR DESCRIPTION
**Pull request by vipulnsward**
Add an option `end` to `find_in_batches`
that complements the `start`parameter to specify where to stop batch processing

### CODE REVIEW

**Comment by rafaelfranca** [comment]
Seems good. Could you update the guides now?

**Comment by rafaelfranca** [comment]
Also needs to update `find_each`

**Comment by vipulnsward** [comment]
@rafaelfranca updated and added documentation for new parameter. 
Needed to change the name from `end` to `end_at`, since we are using keyword arguments now.

**Comment by kaspth** [comment]
You could also consider calling it stop. Usually end goes with begin and stop goes with start.
Currently you're mixing metaphors.

**Comment by vipulnsward** [comment]
stop doesn't sound good to me. Consider this sentence - `The starting value for this query is "start" and stopping value is "stop"`,  here `end` is more suitable, since you are usually meaning `ending` value(PS- I know using 'stopping' doesn't make sense in the sentence here). 
In ruby world begin/end complement each other, in our case they don't.
How does `start_at`/`end_at` sound ?

File: activerecord/lib/active_record/relation/batches.rb [hunk 1]
```diff
@@ -45,14 +46,21 @@ module Batches
...
+        find_in_batches(start: start, end_at: end_at, batch_size: batch_size) do |records|
           records.each { |record| yield record }
         end
       else
-        enum_for(:find_each, start: start, batch_size: batch_size) do
-          start ? where(table[primary_key].gteq(start)).size : size
+        enum_for(:find_each, start: start, batch_size: batch_size, end_at: end_at) do
+          relation = self
+          if start || end_at
+            relation = relation.where(table[primary_key].gteq(start)) if start
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/relation/batches.rb]: I think we can extract this code to a private method. It is duplicated in three places

**Comment by rafaelfranca** [comment]
Renaming `start` will be a very unnecessary breaking change. Maybe `ends_at` sounds better.

**Comment by kaspth** [comment]
Okay, `end_at` (without s) has got my vote.
I do like the parity of `start_at` and `end_at`, but I agree with Rafael the rename isn't worth it.

**Comment by rafaelfranca** [comment]
:+1:

**Comment by vipulnsward** [comment]
@rafaelfranca done. Its much cleaner now

File: activerecord/lib/active_record/relation/batches.rb [hunk 2]
```diff
@@ -28,10 +28,11 @@ module Batches
     # ==== Options
     # * <tt>:batch_size</tt> - Specifies the size of the batch. Default to 1000.
     # * <tt>:start</tt> - Specifies the primary key value to start from.
+    # * <tt>:end_at</tt> - Specifies the ending point for the batch processing.
```
Comment (kaspth) [inline activerecord/lib/active_record/relation/batches.rb]: What does an ending point mean? Let's clarify by mimicking the line above.

"Specifies the primary key value to end at."

File: activerecord/lib/active_record/relation/batches.rb [hunk 3]
```diff
@@ -127,6 +131,12 @@ def find_in_batches(start: nil, batch_size: 1000)
 
     private
 
+    def apply_limits relation, start, end_at
```
Comment (kaspth) [inline activerecord/lib/active_record/relation/batches.rb]: You need parentheses around parameters :smile:

File: activerecord/lib/active_record/relation/batches.rb [hunk 4]
```diff
@@ -78,10 +80,11 @@ def find_each(start: nil, batch_size: 1000)
     # ==== Options
     # * <tt>:batch_size</tt> - Specifies the size of the batch. Default to 1000.
     # * <tt>:start</tt> - Specifies the primary key value to start from.
+    # * <tt>:end_at</tt> - Specifies the ending point for the batch processing.
     # This is especially useful if you want multiple workers dealing with
     # the same processing queue. You can make worker 1 handle all the records
     # between id 0 and 10,000 and worker 2 handle from 10,000 and beyond
-    # (by setting the +:start+ option on that worker).
+    # (by setting the +:start+ and +:end_at+ option on that worker).
```
Comment (matthewd) [inline activerecord/lib/active_record/relation/batches.rb]: `s/that/each/`

File: activerecord/lib/active_record/relation/batches.rb (hunk 2 above)
Comment (matthewd) [inline activerecord/lib/active_record/relation/batches.rb]: Less rhetorically: I just read all the API doc changes, and @kaspth's proposed alternate spelling, and was still unclear on whether it was inclusive or exclusive. I suggest we make that very explicit.

For that matter... have we considered and dismissed the idea of a `:range` option instead?

Comment (kaspth) [inline activerecord/lib/active_record/relation/batches.rb]: :+1: on removing the ambiguity.

No, we haven't discussed a range option. Off the top of my head that sounds less clear re. inclusive/exclusive.

Comment (kaspth) [inline activerecord/lib/active_record/relation/batches.rb]: Actually, I'd be :-1: on having a range option. We can't really change the start option. So the end option is a nice compliment, whereas the range plus start-end seems excessive.

**Comment by vipulnsward** [comment]
Updated. How does this look?

File: activerecord/lib/active_record/relation/batches.rb [hunk 5]
```diff
@@ -45,14 +46,15 @@ module Batches
...
+    def find_each(start: nil, end_at: nil, batch_size: 1000)
       if block_given?
-        find_in_batches(start: start, batch_size: batch_size) do |records|
+        find_in_batches(start: start, end_at: end_at, batch_size: batch_size) do |records|
           records.each { |record| yield record }
         end
       else
-        enum_for(:find_each, start: start, batch_size: batch_size) do
-          start ? where(table[primary_key].gteq(start)).size : size
+        enum_for(:find_each, start: start, batch_size: batch_size, end_at: end_at) do
```
Comment (kaspth) [inline activerecord/lib/active_record/relation/batches.rb]: Let's insert `end_at` before `batch_size` as above for consistency (it's nice that start and end is together) :smile:

File: activerecord/lib/active_record/relation/batches.rb [hunk 6]
```diff
@@ -95,12 +98,12 @@ def find_each(start: nil, batch_size: 1000)
...
     # NOTE: You can't set the limit either, that's used to control
     # the batch sizes.
-    def find_in_batches(start: nil, batch_size: 1000)
+    def find_in_batches(start: nil, end_at: nil, batch_size: 1000)
       relation = self
 
       unless block_given?
-        return to_enum(:find_in_batches, start: start, batch_size: batch_size) do
-          total = start ? where(table[primary_key].gteq(start)).size : size
+        return to_enum(:find_in_batches, start: start, batch_size: batch_size, end_at: end_at) do
```
Comment (kaspth) [inline activerecord/lib/active_record/relation/batches.rb]: Same here.

File: guides/source/active_record_querying.md [hunk 7]
```diff
@@ -343,6 +343,19 @@ end
 
 Another example would be if you wanted multiple workers handling the same processing queue. You could have each worker handle 10000 records by setting the appropriate `:start` option on each worker.
 
+**`:end_at`**
+
+Similar to `:start` option, `:end_at` allows you to configure the last ID of the sequence whenever the highest ID is not the one you need.
```
Comment (kaspth) [inline guides/source/active_record_querying.md]: Similar to *the* `:start` option

**Comment by kaspth** [comment]
Some very minor things and then it LGTM. Thanks, Vipul :heart:

**Comment by vipulnsward** [comment]
Thanks @kaspth . Updated again.

**Comment by dhh** [comment]
I'd like to bring parity to the options. Either it's start/stop or begin_at/end_at. But mixing the language like now isn't great. I think it's fine to do a deprecation on start to go with begin_at/end_at. We don't want a mixed language wart sitting in the code base forever.

**Comment by vipulnsward** [comment]
@dhh got it. Will send a PR to take care of this.

**Comment by bogdan** [comment]
In my head `_at` is associated with a timestamp but not id. So `begin_at end_at` looks like a date range.

**Comment by piotrj** [comment]
@bogdan My first thought was "what are the dates doing in find_in_batches" as well

**Comment by anderslemke** [comment]
I agree that the `_at` implies date/time ranges.

Maybe `first` and `last` might work?

**Comment by dhh** [comment]
We can also use start/finish. Neither of those are keywords. And agree with the _at implication of date.

**Comment by vipulnsward** [comment]
If it looks good to everyone, I l change it back to `start`/`finish` ?

**Comment by kaspth** [comment]
@vipulnsward Rock our worlds from start to finish :metal:

**Pull request merged**